File: leetcode/254. Factor Combinations/254.cpp
// https://helloacm.com/algorithms-to-compute-the-factor-combinations-for-an-integer-using-dfs-and-bfs/
// https://leetcode.com/problems/factor-combinations/

class Solution {
public:
    vector<vector<int>> getFactors(int n) {
        dfs(n, {});        
        // remove the [n] factor
        if (!r.empty()) r.erase(end(r), end(r) + 1);
        return r;
    }
private:
    void dfs(int n, vector<int> cur) {
        if (n == 1) {
            if (!cur.empty()) r.push_back(cur);
            return;
        }
        int k = 2;
        if (!cur.empty()) {
            k = max(k, cur.back());
        }
        for (int i = k; i <= n; ++ i) {
            if ((n % i) == 0) { // current is a factor
                cur.push_back(i);
                dfs(n / i, cur); 
                cur.pop_back();
            }
        }
    }
    vector<vector<int>> r;
};

File: leetcode/254. Factor Combinations/254.py
# https://helloacm.com/teaching-kids-programming-backtracking-algorithm-to-find-factor-combinations-math-recursive-depth-first-search/
# https://leetcode.com/problems/factor-combinations/
# MEDIUM, DFS

class Solution:
    def getFactors(self, n: int) -> List[List[int]]:
        ans = []
        
        def dfs(n, s, c):
            for i in range(s, int(n**0.5) + 1):
                if n % i == 0:
                    c += [i]
                    ans.append(c[:] + [n // i])
                    dfs(n // i, i, c)
                    c.pop()
        
        dfs(n, 2, [])
        return ans
